File: Tests/Tests/STPPaymentMethodOXXOParamsTests.m
//
//  STPPaymentMethodOXXOParamsTests.m
//  StripeiOS Tests
//
//  Created by Polo Li on 6/16/20.
//  Copyright © 2020 Stripe, Inc. All rights reserved.
//

#import <XCTest/XCTest.h>

#import "STPTestingAPIClient.h"

@interface STPPaymentMethodOXXOParamsTests : XCTestCase

@end

@implementation STPPaymentMethodOXXOParamsTests

- (void)testCreateOXXOPaymentMethod {
    STPAPIClient *client = [[STPAPIClient alloc] initWithPublishableKey:STPTestingDefaultPublishableKey];
    STPPaymentMethodOXXOParams *oxxoParams = [STPPaymentMethodOXXOParams new];

    STPPaymentMethodBillingDetails *billingDetails = [STPPaymentMethodBillingDetails new];
    billingDetails.name = @"Jane Doe";
    billingDetails.email = @"test@test.com";

    STPPaymentMethodParams *params = [STPPaymentMethodParams paramsWithOXXO:oxxoParams
                                                             billingDetails:billingDetails
                                                                   metadata:@{@"test_key": @"test_value"}];

    XCTestExpectation *expectation = [self expectationWithDescription:@"Payment Method OXXO create"];

    [client createPaymentMethodWithParams:params
                               completion:^(STPPaymentMethod * _Nullable paymentMethod, NSError * _Nullable error) {
        [expectation fulfill];

        XCTAssertNil(error, @"Unexpected error creating OXXO PaymentMethod: %@", error);
        XCTAssertNotNil(paymentMethod, @"Failed to create OXXO PaymentMethod");
        XCTAssertNotNil(paymentMethod.stripeId, @"Missing stripeId");
        XCTAssertNotNil(paymentMethod.created, @"Missing created");
        XCTAssertFalse(paymentMethod.liveMode, @"Incorrect livemode");
        XCTAssertEqual(paymentMethod.type, STPPaymentMethodTypeOXXO, @"Incorrect PaymentMethod type");

        // Billing Details
        XCTAssertEqualObjects(paymentMethod.billingDetails.name, @"Jane Doe", @"Incorrect name");
        XCTAssertEqualObjects(paymentMethod.billingDetails.email, @"test@test.com", @"Incorrect email");

        // OXXO Details
        XCTAssertNotNil(paymentMethod.oxxo, @"Missing OXXO");
    }];

    [self waitForExpectationsWithTimeout:STPTestingNetworkRequestTimeout handler:nil];
}

@end

File: Stripe/StripeiOSTests/STPPaymentMethodOXXOParamsTests.swift
//  Converted to Swift 5.8.1 by Swiftify v5.8.28463 - https://swiftify.com/
//
//  STPPaymentMethodOXXOParamsTests.m
//  StripeiOS Tests
//
//  Created by Polo Li on 6/16/20.
//  Copyright © 2020 Stripe, Inc. All rights reserved.
//

import StripeCore
import StripeCoreTestUtils
import StripePaymentsTestUtils

class STPPaymentMethodOXXOParamsTests: STPNetworkStubbingTestCase {
    func testCreateOXXOPaymentMethod() {
        let client = STPAPIClient(publishableKey: STPTestingDefaultPublishableKey)
        let oxxoParams = STPPaymentMethodOXXOParams()

        let billingDetails = STPPaymentMethodBillingDetails()
        billingDetails.name = "Jane Doe"
        billingDetails.email = "test@test.com"

        let params = STPPaymentMethodParams(
            oxxo: oxxoParams,
            billingDetails: billingDetails,
            metadata: [
                "test_key": "test_value"
            ])

        let expectation = self.expectation(description: "Payment Method OXXO create")

        client.createPaymentMethod(
            with: params) { paymentMethod, error in
            expectation.fulfill()

            XCTAssertNil(error, "Unexpected error creating OXXO PaymentMethod")
            XCTAssertNotNil(paymentMethod, "Failed to create OXXO PaymentMethod")
            XCTAssertNotNil(paymentMethod?.stripeId, "Missing stripeId")
            XCTAssertNotNil(paymentMethod?.created, "Missing created")
            XCTAssertFalse(paymentMethod!.liveMode, "Incorrect livemode")
            XCTAssertEqual(paymentMethod?.type, .OXXO, "Incorrect PaymentMethod type")

            // Billing Details
                XCTAssertEqual(paymentMethod?.billingDetails?.name, "Jane Doe")
            XCTAssertEqual(paymentMethod?.billingDetails?.email, "test@test.com")

            // OXXO Details
            XCTAssertNotNil(paymentMethod?.oxxo, "Missing OXXO")
        }

        waitForExpectations(timeout: STPTestingNetworkRequestTimeout, handler: nil)
    }
}
